cpp/src/jni/orc/jni_wrapper.cpp

```diff
@@ -0,0 +1,296 @@
+  return reader->NumberOfStripes();
+}
+
+JNIEXPORT jlong JNICALL
+Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_nextStripeReader(JNIEnv* env,
+                                                                       jobject this_obj,
+                                                                       jlong id,
+                                                                       jlong batch_size) {
+  auto reader = orc_reader_holder_.Lookup(id);
+  if (!reader) {
+    std::string error_message = "invalid reader id " + std::to_string(id);
+    env->ThrowNew(exception_class, error_message.c_str());
+  }
+
+  std::shared_ptr<RecordBatchReader> stripe_reader;
+  auto status = reader->NextStripeReader(batch_size, &stripe_reader);
+
+  if (!status.ok()) {
+    return static_cast<jlong>(status.code()) * -1;
+  }
+
+  if (!stripe_reader) {
+    return static_cast<jlong>(arrow::StatusCode::Invalid) * -1;
+  }
+
+  return orc_stripe_reader_holder_.Insert(stripe_reader);
+}
+
+JNIEXPORT jbyteArray JNICALL
+Java_org_apache_arrow_adapter_orc_OrcStripeReaderJniWrapper_getSchema(JNIEnv* env,
+                                                                      jclass this_cls,
+                                                                      jlong id) {
+  auto stripe_reader = orc_stripe_reader_holder_.Lookup(id);
+  if (!stripe_reader) {
+    std::string error_message = "invalid stripe reader id " + std::to_string(id);
+    env->ThrowNew(exception_class, error_message.c_str());
+  }
+
+  auto schema = stripe_reader->schema();
+
+  std::shared_ptr<arrow::Buffer> out;
+  auto status =
+      arrow::ipc::SerializeSchema(*schema, nullptr, arrow::default_memory_pool(), &out);
+  if (!status.ok()) {
+    return nullptr;
+  }
+
+  jbyteArray ret = env->NewByteArray(out->size());
+  auto src = reinterpret_cast<const jbyte*>(out->data());
+  env->SetByteArrayRegion(ret, 0, out->size(), src);
+  return ret;
+}
+
+JNIEXPORT jobject JNICALL
+Java_org_apache_arrow_adapter_orc_OrcStripeReaderJniWrapper_next(JNIEnv* env,
+                                                                 jclass this_cls,
+                                                                 jlong id) {
+  auto stripe_reader = orc_stripe_reader_holder_.Lookup(id);
+  if (!stripe_reader) {
+    std::string error_message = "invalid stripe reader id " + std::to_string(id);
+    env->ThrowNew(exception_class, error_message.c_str());
+  }
+
+  std::shared_ptr<arrow::RecordBatch> record_batch;
+  auto status = stripe_reader->ReadNext(&record_batch);
+  if (!status.ok() || !record_batch) {
+    return nullptr;
+  }
+
+  auto schema = stripe_reader->schema();
+
+  // create OrcFieldNode[]
+  jobjectArray field_array =
+      env->NewObjectArray(schema->num_fields(), orc_field_node_class, nullptr);
+
+  std::vector<std::shared_ptr<arrow::Buffer>> buffers;
+  for (int i = 0; i < schema->num_fields(); ++i) {
+    auto column = record_batch->column(i);
```

Comment:
there are a lot of local references being created to java objects - can you please test with a large file say of 100 columns or more and see if this goes through..from what i remember jvm needed native methods to inform it if more than 16 references were needed using ensureLocalCapacity()..

Reference : https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/functions.html

I am ok if you want to create a followup JIRA for this.

Reply:
https://issues.apache.org/jira/browse/ARROW-5515